File: services/crawler-titan/core/workset.py
"""竞彩 workset 状态 + 比赛明细文件（镜像 sporttery src/sources/jingcai/workset.ts）。

workset.json 结构（以 business_date 分组）：
  {
    "version": 3,
    "updatedAt": "...",
    "completeDate": "2026-08-01",
    "dates": {
      "2026-08-11": {
        "attempts": 0,
        "last_settle_at": null,
        "matches": { "<sid>": {sid, business_date, kickoff, status, match_num,
                               sclass_id, sub_id, home/away team ids+names, scores,
                               analysis_done, odds_done, last_odds_at, odds_phase} }
      }
    }
  }

比赛明细文件 data/jc/matches/{sid}.json：
  {
    "match_id": sid, "business_date": "...", "kickoff_time": "...", "last_odds_at": "...",
    "detail": {
      "analysis": {"matches": {...}, "h2h": [...], "recent_home": [...], "recent_away": [...]},
      "odds": {"asian": [...], "over_under": [...], "european": [...]}   # 每类 = 最新全量 changes 列表
    }
  }

旧平铺格式 {"matches": {sid: {...}}} 自动迁移为 dates 结构。
"""
import json
import os
import time

from core import utils

WORKSET_PATH = os.path.join(utils.DATA_DIR, "jc", "workset.json")
MATCHES_DIR = os.path.join(utils.DATA_DIR, "jc", "matches")

VERSION = 3


class Workset:
    def __init__(self):
        self.data = {
            "version": VERSION,
            "updatedAt": "",
            "completeDate": None,
            "dates": {},
        }

    # ─── 持久化 ──────────────────────────────────────────────

    def load(self) -> None:
        if os.path.isfile(WORKSET_PATH):
            try:
                with open(WORKSET_PATH, "r", encoding="utf-8") as f:
                    parsed = json.load(f)
                self.data = {
                    "version": VERSION,
                    "updatedAt": parsed.get("updatedAt", ""),
                    "completeDate": parsed.get("completeDate"),
                    "dates": parsed.get("dates", {}),
                    "matches": parsed.get("matches", {}),   # 旧平铺格式，供 _migrate 转换
                }
                self._migrate()
                return
            except (json.JSONDecodeError, OSError):
                pass
        self.data = {"version": VERSION, "updatedAt": "", "completeDate": None, "dates": {}}

    def save(self) -> None:
        self.data["updatedAt"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        os.makedirs(os.path.dirname(WORKSET_PATH), exist_ok=True)
        tmp = f"{WORKSET_PATH}.tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(self.data, f, ensure_ascii=False, indent=2)
        os.replace(tmp, WORKSET_PATH)

    def _migrate(self) -> None:
        """旧平铺 {"matches": {sid: {...}}} → dates 分组结构。"""
        matches = self.data.get("matches")
        if not isinstance(matches, dict) or not matches:
            return
        for sid, m in matches.items():
            date = m.get("business_date")
            if not date:
                continue
            entry = self.data["dates"].setdefault(date, {"attempts": 0, "last_settle_at": None, "matches": {}})
            entry["matches"][str(sid)] = m
        self.data.pop("matches", None)

    # ─── completeDate ────────────────────────────────────────

    @property
    def complete_date(self):
        return self.data.get("completeDate")

    def set_complete_date(self, date: str) -> None:
        self.data["completeDate"] = date

    # ─── dates / matches ─────────────────────────────────────

    def dates(self) -> list:
        return sorted(self.data["dates"].keys())

    def total_matches(self) -> int:
        return sum(len(entry.get("matches", {})) for entry in self.data["dates"].values())

    def matches_of(self, date: str) -> dict:
        return self.data["dates"].get(date, {}).get("matches", {})

    def date_entry(self, date: str) -> dict:
        return self.data["dates"].setdefault(date, {"attempts": 0, "last_settle_at": None, "matches": {}})

    def remove_date(self, date: str) -> None:
        self.data["dates"].pop(date, None)

    def attempts(self, date: str) -> int:
        return self.date_entry(date).get("attempts", 0)

    def increment_attempts(self, date: str) -> None:
        self.date_entry(date)["attempts"] = self.attempts(date) + 1

    def last_settle_at(self, date: str):
        return self.date_entry(date).get("last_settle_at")

    def set_last_settle_at(self, date: str, iso: str) -> None:
        self.date_entry(date)["last_settle_at"] = iso

    def prune_empty_dates(self) -> None:
        for date in list(self.data["dates"].keys()):
            if not self.data["dates"][date].get("matches"):
                del self.data["dates"][date]

    # ─── upsert match ────────────────────────────────────────

    def upsert_match(self, m: dict) -> None:
        """合并/新增一条比赛（来自 bf_jc / JcResult 解析行）。"""
        sid = m.get("sid")
        date = m.get("business_date")
        if sid is None or not date:
            return
        entry = self.date_entry(date)
        key = str(sid)
        prev = entry["matches"].get(key, {})
        merged = dict(prev)
        for k, v in m.items():
            if v is not None or k not in merged:
                merged[k] = v
        merged.setdefault("analysis_done", prev.get("analysis_done", False))
        merged.setdefault("odds_done", prev.get("odds_done", False))
        merged.setdefault("first_odds_at", prev.get("first_odds_at"))
        merged.setdefault("last_odds_at", prev.get("last_odds_at"))
        merged.setdefault("final_odds_fetched", prev.get("final_odds_fetched", False))
        merged.setdefault("refund_checked", prev.get("refund_checked", False))
        merged.setdefault("backup_done", prev.get("backup_done", False))
        merged.setdefault("odds_phase", prev.get("odds_phase", "pending"))
        merged["sid"] = sid
        merged["business_date"] = date
        entry["matches"][key] = merged

    # ─── 比赛明细文件 ────────────────────────────────────────

    def match_file_path(self, sid) -> str:
        return os.path.join(MATCHES_DIR, f"{sid}.json")

    def match_file_exists(self, sid) -> bool:
        return os.path.isfile(self.match_file_path(sid))

    def new_match_file(self, sid, business_date, kickoff=None) -> dict:
        return {
            "match_id": sid,
            "business_date": business_date,
            "kickoff_time": kickoff,
            "last_odds_at": None,
            "detail": {},
        }

    def read_match(self, sid) -> dict:
        p = self.match_file_path(sid)
        if not os.path.isfile(p):
            return None
        try:
            with open(p, "r", encoding="utf-8") as f:
                return json.load(f)
        except (json.JSONDecodeError, OSError):
            return None

    def save_match(self, jf: dict) -> None:
        os.makedirs(MATCHES_DIR, exist_ok=True)
        tmp = f"{self.match_file_path(jf['match_id'])}.tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(jf, f, ensure_ascii=False, indent=2)
        os.replace(tmp, self.match_file_path(jf["match_id"]))

    def delete_match_files(self, date: str) -> None:
        for m in self.matches_of(date).values():
            p = self.match_file_path(m.get("sid"))
            if os.path.isfile(p):
                try:
                    os.unlink(p)
                except OSError:
                    pass

    # ─── 就绪判定 ────────────────────────────────────────────

    def is_date_ready(self, date: str) -> bool:
        """整日就绪 = 所有比赛已完赛 且 每场都有完整 detail（analysis + odds）。"""
        matches = self.matches_of(date)
        if not matches:
            return False
        for m in matches.values():
            if not is_terminal(m):
                return False
            jf = self.read_match(m.get("sid"))
            if not jf:
                return False
            detail = jf.get("detail") or {}
            if not detail.get("analysis") or not detail.get("odds"):
                return False
        return True


def is_settled(m: dict) -> bool:
    return m.get("status") == -1


def is_terminal(m: dict) -> bool:
    """完场（-1）或异常终止（-10）。"""
    return m.get("status") in (-1, -10)

File: services/crawler-sporttery/src/sources/jingcai/workset.ts
import * as fs from "fs";
import * as path from "path";
import { isRefund, isSettled } from "./completeness.js";

const WORKSET_PATH = path.join(import.meta.dirname, "../../../data/jingcai/workset.json");
const MATCHES_DIR = path.join(path.dirname(WORKSET_PATH), "matches");

export interface WorksetMatch {
  matchId: number;
  businessDate: string;
  matchDate?: string;
  matchNum?: string;
  homeTeam?: string;
  homeTeamId?: number;
  awayTeam?: string;
  awayTeamId?: number;
  league?: string;
  leagueId?: number;
  kickoffTime?: string | null;
  matchResult?: string;
  poolStatus?: string;
  had?: any;
  handicap?: any;
  firstOddsAt?: string | null;
  lastOddsAt?: string | null;
  finalOddsFetched?: boolean;
  detailFetched?: boolean;
  [key: string]: any;
}

export interface VoteSnapshot {
  at: string;
  had?: any;
  handicap?: any;
}

export interface MatchFile {
  matchId: number;
  businessDate: string;
  kickoffTime?: string | null;
  detail?: any;
  voteSnapshots?: VoteSnapshot[];
  lastOddsAt?: string;
  lastVoteAt?: string;
}

interface DateEntry {
  attempts: number;
  lastVoteAt?: string | null;
  matches: WorksetMatch[];
}

interface WorksetData {
  version: number;
  updatedAt: string;
  completeDate: string | null;
  dates: Record<string, DateEntry>;
}

export class Workset {
  private data: WorksetData = { version: 3, updatedAt: "", completeDate: null, dates: {} };

  load(): void {
    if (fs.existsSync(WORKSET_PATH)) {
      try {
        const parsed = JSON.parse(fs.readFileSync(WORKSET_PATH, "utf-8"));
        this.data = {
          version: 3,
          updatedAt: parsed?.updatedAt ?? "",
          completeDate: parsed?.completeDate ?? null,
          dates: parsed?.dates ?? {},
        };
        this.migrate();
      } catch {
        this.data = { version: 3, updatedAt: "", completeDate: null, dates: {} };
      }
    }
  }

  // v2 → v3：把旧 workset 里内嵌的 detail 迁移到 matches/{id}.json
  private migrate(): void {
    for (const date of Object.keys(this.data.dates)) {
      const entry = this.data.dates[date];
      for (const m of entry.matches) {
        if (m.detail) {
          const jf = this.readMatch(m.matchId) ?? this.newMatchFile(m.matchId, m.businessDate);
          if (!jf.detail) {
            jf.detail = m.detail;
            jf.kickoffTime = m.detail?.matchInfo?.matchDateTime ?? m.kickoffTime ?? null;
          }
          jf.voteSnapshots = jf.voteSnapshots ?? [];
          this.saveMatch(jf);
          m.detailFetched = true;
          m.kickoffTime = m.kickoffTime ?? jf.kickoffTime ?? null;
          delete m.detail;
        }
      }
    }
  }

  get completeDate(): string | null {
    return this.data.completeDate;
  }

  setCompleteDate(date: string): void {
    this.data.completeDate = date;
  }

  get dates(): string[] {
    return Object.keys(this.data.dates).sort();
  }

  hasDate(date: string): boolean {
    return date in this.data.dates;
  }

  matchesOf(date: string): WorksetMatch[] {
    return this.data.dates[date]?.matches ?? [];
  }

  dateEntry(date: string): DateEntry {
    this.ensureDate(date);
    return this.data.dates[date];
  }

  lastVoteAt(date: string): string | null {
    return this.data.dates[date]?.lastVoteAt ?? null;
  }

  setLastVoteAt(date: string, iso: string): void {
    this.ensureDate(date);
    this.data.dates[date].lastVoteAt = iso;
  }

  attempts(date: string): number {
    return this.data.dates[date]?.attempts ?? 0;
  }

  ensureDate(date: string): void {
    if (!this.data.dates[date]) this.data.dates[date] = { attempts: 0, matches: [] };
  }

  incrementAttempts(date: string): void {
    this.ensureDate(date);
    this.data.dates[date].attempts++;
  }

  isEmpty(date: string): boolean {
    return (this.data.dates[date]?.matches?.length ?? 0) === 0;
  }

  // workset 里全部比赛数（=0 即进入 DISCOVERY 态）
  totalMatches(): number {
    let n = 0;
    for (const date of Object.keys(this.data.dates)) {
      n += this.data.dates[date]?.matches?.length ?? 0;
    }
    return n;
  }

  // 清理空条目（空条目无任何意义：发现/抓取命中时会自然创建带比赛的条目）
  pruneEmptyDates(): void {
    for (const date of Object.keys(this.data.dates)) {
      const len = this.data.dates[date]?.matches?.length ?? 0;
      if (len === 0) delete this.data.dates[date];
    }
  }

  isDateReady(date: string): boolean {
    const matches = this.matchesOf(date);
    return matches.length > 0 && matches.every((m) => isSettled(m) || isRefund(m));
  }

  removeDate(date: string): void {
    delete this.data.dates[date];
  }

  // 合并每日 getVoteV1 元数据；不处理 detail（由调用方落 JSON）
  upsertMatches(matches: any[]): void {
    for (const m of matches) {
      const date = m?.businessDate;
      if (!date || !m?.matchId) continue;
      this.ensureDate(date);
      const entry = this.data.dates[date];
      const idx = entry.matches.findIndex((x) => x.matchId === m.matchId);
      const incoming = { ...m };
      delete incoming.detail;
      if (idx >= 0) {
        const prev = entry.matches[idx];
        entry.matches[idx] = {
          ...prev,
          ...incoming,
          kickoffTime: prev.kickoffTime ?? incoming.kickoffTime ?? null,
          firstOddsAt: prev.firstOddsAt ?? incoming.firstOddsAt ?? null,
          lastOddsAt: prev.lastOddsAt ?? incoming.lastOddsAt ?? null,
          finalOddsFetched: prev.finalOddsFetched ?? incoming.finalOddsFetched ?? false,
          detailFetched: prev.detailFetched ?? incoming.detailFetched ?? false,
        };
      } else {
        entry.matches.push({
          ...incoming,
          kickoffTime: incoming.kickoffTime ?? null,
          firstOddsAt: null,
          lastOddsAt: null,
          finalOddsFetched: false,
          detailFetched: false,
        });
      }
    }
  }

  // ─── matches/{id}.json 文件操作 ───────────────────────────
  matchFilePath(matchId: number): string {
    return path.join(MATCHES_DIR, `${matchId}.json`);
  }

  matchFileExists(matchId: number): boolean {
    return fs.existsSync(this.matchFilePath(matchId));
  }

  newMatchFile(matchId: number, businessDate: string): MatchFile {
    return { matchId, businessDate, kickoffTime: null, detail: undefined, voteSnapshots: [], lastOddsAt: undefined, lastVoteAt: undefined };
  }

  readMatch(matchId: number): MatchFile | null {
    const p = this.matchFilePath(matchId);
    if (!fs.existsSync(p)) return null;
    try {
      return JSON.parse(fs.readFileSync(p, "utf-8")) as MatchFile;
    } catch {
      return null;
    }
  }

  saveMatch(jf: MatchFile): void {
    fs.mkdirSync(MATCHES_DIR, { recursive: true });
    const tmp = `${this.matchFilePath(jf.matchId)}.tmp`;
    fs.writeFileSync(tmp, JSON.stringify(jf, null, 2), "utf-8");
    fs.renameSync(tmp, this.matchFilePath(jf.matchId));
  }

  // 排干成功后删除该日所有比赛的 JSON
  deleteMatchFiles(date: string): void {
    for (const m of this.matchesOf(date)) {
      const p = this.matchFilePath(m.matchId);
      if (fs.existsSync(p)) {
        try { fs.unlinkSync(p); } catch {}
      }
    }
  }

  // 追加一条 votes 快照（与上一快照数值一致则不落），返回是否落盘
  recordVoteSnapshot(matchId: number, businessDate: string, at: string, had?: any, handicap?: any): boolean {
    let jf = this.readMatch(matchId);
    if (!jf) {
      jf = this.newMatchFile(matchId, businessDate);
    }
    jf.voteSnapshots = jf.voteSnapshots ?? [];
    const last = jf.voteSnapshots[jf.voteSnapshots.length - 1];
    if (last && last.at === at && jsonEqual(last.had, had) && jsonEqual(last.handicap, handicap)) {
      return false;
    }
    jf.voteSnapshots.push({ at, had, handicap });
    jf.lastVoteAt = at;
    this.saveMatch(jf);
    return true;
  }

  save(): void {
    this.data.updatedAt = new Date().toISOString();
    const tmp = `${WORKSET_PATH}.tmp`;
    fs.mkdirSync(path.dirname(WORKSET_PATH), { recursive: true });
    fs.writeFileSync(tmp, JSON.stringify(this.data, null, 2), "utf-8");
    fs.renameSync(tmp, WORKSET_PATH);
  }
}

function jsonEqual(a: unknown, b: unknown): boolean {
  return JSON.stringify(a ?? null) === JSON.stringify(b ?? null);
}
